src/node/txdownload_impl.cpp

```diff
@@ -0,0 +1,455 @@
+// Copyright (c) 2023
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <node/txdownload_impl.h>
+
+#include <chain.h>
+#include <consensus/validation.h>
+#include <logging.h>
+#include <txmempool.h>
+#include <validation.h>
+#include <validationinterface.h>
+
+namespace node {
+void TxDownloadImpl::UpdatedBlockTipSync()
+{
+    // If the chain tip has changed, previously rejected transactions might now be invalid, e.g. due
+    // to a timelock. Reset the rejection filters to give those transactions another chance if we
+    // see them again.
+    m_recent_rejects.reset();
+    m_recent_rejects_reconsiderable.reset();
+}
+
+void TxDownloadImpl::BlockConnected(const std::shared_ptr<const CBlock>& pblock)
+{
+    m_orphanage.EraseForBlock(*pblock);
+
+    for (const auto& ptx : pblock->vtx) {
+        m_recent_confirmed_transactions.insert(ptx->GetHash().ToUint256());
+        if (ptx->HasWitness()) {
+            m_recent_confirmed_transactions.insert(ptx->GetWitnessHash().ToUint256());
+        }
+    }
+    for (const auto& ptx : pblock->vtx) {
+        m_txrequest.ForgetTxHash(ptx->GetHash());
+        m_txrequest.ForgetTxHash(ptx->GetWitnessHash());
+    }
+}
+
+void TxDownloadImpl::BlockDisconnected()
+{
+    // To avoid relay problems with transactions that were previously
+    // confirmed, clear our filter of recently confirmed transactions whenever
+    // there's a reorg.
+    // This means that in a 1-block reorg (where 1 block is disconnected and
+    // then another block reconnected), our filter will drop to having only one
+    // block's worth of transactions in it, but that should be fine, since
+    // presumably the most common case of relaying a confirmed transaction
+    // should be just after a new block containing it is found.
+    m_recent_confirmed_transactions.reset();
+}
+
+bool TxDownloadImpl::AlreadyHaveTx(const GenTxid& gtxid, bool include_reconsiderable)
+{
+    const uint256& hash = gtxid.GetHash();
```

Comment:
Iiuc correctly, clearing the filters here has been removed as we don't want `TxDownloadManager` to depend on `ChainstateManager`.

An alternative to the new synchronous validation interface callback could be to preserve the previous behavior (clearing the filters in `AlreadyHaveTx` on a tip change) by introducing an interface for TxDownloadMan to fetch the latest tip hash. For example, `TxDownloadOptions` could be extended to hold a lambda that returns the latest tip hash, which in production simply fetches the hash from the `ChainstateManager` (and for tests it can be mocked). This might be easier to review since it preserves the previous mechanism? Although perhaps also more difficult w.r.t to lock inversion of `cs_main` and `m_tx_download_mutex`.

Another alternative could be to directly pass the current block hash as a param into the relevant `TxDownloadManager` methods.

Reply:
I don't think we'd ever need to make `TxDownloadManager` depend on `ChainstateManager` as we could just pass in the block hash, i.e.

> Another alternative could be to directly pass the current block hash as a param into the relevant TxDownloadManager methods.

This would avoid the addition of `UpdateBlockTipSync` etc, as it's basically what we do now. I find this way of keeping synchronized with the chain tip pretty ugly - we'd need to hold `cs_main` and pass the blockhash to the `TxDownloadManager` all the time, bloating the interface. Subscribing to the validation interface and updating on every chain tip update seems much more sensible.